dmq-node/src/DMQ/Protocol/SigSubmission/Validate.hs

```diff
@@ -0,0 +1,199 @@
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Except
+import Control.Monad.Trans.Except.Extra
+import Data.Aeson
+import Data.ByteString (ByteString)
+import Data.ByteString.Lazy qualified as LBS
+import Data.Map.Strict qualified as Map
+import Data.Maybe (fromJust, isNothing)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Typeable
+import Data.Word
+
+import Cardano.Crypto.DSIGN.Class (ContextDSIGN)
+import Cardano.Crypto.DSIGN.Class qualified as DSIGN
+import Cardano.Crypto.KES.Class (KESAlgorithm (..))
+import Cardano.KESAgent.KES.Crypto as KES
+import Cardano.KESAgent.KES.OCert (OCert (..), OCertSignable, validateOCert)
+import Cardano.Ledger.BaseTypes.NonZero
+import Cardano.Ledger.Hashes
+
+import DMQ.Diffusion.NodeKernel (PoolValidationCtx (..))
+import DMQ.Protocol.SigSubmission.Type
+import Ouroboros.Consensus.Shelley.Ledger.Query
+import Ouroboros.Network.TxSubmission.Mempool.Simple
+import Ouroboros.Network.Util.ShowProxy
+
+
+-- | The type of non-fatal failures reported by the mempool writer
+-- for invalid messages
+--
+data instance TxValidationFail (Sig crypto) =
+    SigInvalid SigValidationError
+  | SigDuplicate
+  | SigExpired
+  | SigResultOther Text
+  deriving (Eq, Show)
+
+instance (Typeable crypto) => ShowProxy (TxValidationFail (Sig crypto))
+
+instance ToJSON (TxValidationFail (Sig crypto)) where
+  toJSON SigDuplicate = String "duplicate"
+  toJSON SigExpired   = String "expired"
+  toJSON (SigInvalid e) = object
+    [ "type" .= String "invalid"
+    , "reason" .= show e
+    ]
+  toJSON (SigResultOther txt) = object
+    [ "type" .= String "other"
+    , "reason" .= txt
+      ]
+
+
+data SigValidationError =
+    InvalidKESSignature KESPeriod KESPeriod String
+  | InvalidSignatureOCERT
+      !Word64    -- OCert counter
+      !KESPeriod -- OCert KES period
+      !String    -- DSIGN error message
+  | InvalidOCertCounter
+      Word64 -- last seen
+      Word64 -- received
+  | KESBeforeStartOCERT KESPeriod KESPeriod
+  | KESAfterEndOCERT KESPeriod KESPeriod
+  | UnrecognizedPool
+  | NotInitialized
+  | ClockSkew
+  deriving (Eq, Show)
+
+
+-- TODO:
+--  We don't validate ocert numbers, since we might not have necessary
+--  information to do so, but we can validate that they are growing.
+validateSig :: forall crypto m.
+               ( Crypto crypto
+               , ContextDSIGN (KES.DSIGN crypto) ~ ()
+               , DSIGN.Signable (DSIGN crypto) (OCertSignable crypto)
+               , ContextKES (KES crypto) ~ ()
+               , Signable (KES crypto) ByteString
+               , MonadSTM m
+               )
+            => (DSIGN.VerKeyDSIGN (DSIGN crypto) -> KeyHash StakePool)
+            -> [Sig crypto]
+            -> PoolValidationCtx m
+            -- ^ cardano pool id verification
+            -> ExceptT (Sig crypto, TxValidationFail (Sig crypto)) m
+                       [(Sig crypto, Either (TxValidationFail (Sig crypto)) ())]
```

Comment:
Why some errors are return on in the `ExceptT` while others in the result of the validation?